factory/README.md

```diff
@@ -0,0 +1,201 @@
+# cypress/factory
+
+`cypress/factory` is a docker image that can be used with docker [`args`](https://docs.docker.com/engine/reference/builder/#arg) to generate a docker container with specific versions of:
+
+* node
+* yarn
+* chrome
+* firefox
+* edge
+* cypress
+
+## Benefits
+
+* Freedom to choose which versions to test against.
+* No need to wait on an official release to test the latest version of a browser.
+* Smaller docker sizes especially when not including unused browsers.
+* Easily test multiple browser versions.
+* Reduced maintenance and pull requests for the cypress-docker repo.
+* Ability to offer more variations of docker containers at low cost.
+
+## API
+
+The following args can be set to define what versions the cypress factory docker includes in it's final build.
+
+If no args are defined, only the default version of node will be installed. This can still be a useful container though since we will also install any required cypress apt dependencies.
+
+### NODE_VERSION
+
+The version of Node to install in the docker container. If not specified, the default version of Node (defined [here](./.env)) is installed. Node is required. The exact version must be used, no wildcards or shorthands are supported.
+
+Example: `NODE_VERSION='16.18.1'`
+
+[Node Versions](https://nodejs.org/en/download/releases)
+
+### YARN_VERSION
+
+The version of yarn to install (via npm). If not specified, Yarn is not installed.
+
+Example: `YARN_VERSION='1.22.19'`
+
+[Yarn versions](https://www.npmjs.com/package/yarn)
+
+### CYPRESS_VERSION
+
+The version of Cypress to install (via npm). If not specified, Cypress is not installed.
+
+Example: `CYPRESS_VERSION='12.1.0'`
+
+[Cypress versions](https://www.npmjs.com/package/cypress)
+
+### CHROME_VERSION
+
+The version of Chrome to install. If not specified, Chrome is not installed. The exact version must be used, no wildcards or shorthands are supported.
+
+Example: `CHROME_VERSION='107.0.5304.121-1'`
+
+[Chrome versions](https://www.ubuntuupdates.org/package/google_chrome/stable/main/base/google-chrome-stable)
+
+### FIREFOX_VERSION
+
+The version of Firefox to install. If not specified, Firefox is not installed. The exact version must be used, no wildcards or shorthands are supported.
+
+Example: `FIREFOX_VERSION='107.0'`
+
+[Firefox versions](https://download-installer.cdn.mozilla.net/pub/firefox/releases/)
+
+### EDGE_VERSION
+
+The version of Edge to install. If not specified, Edge is not installed. The exact version must be used, no wildcards or shorthands are supported.
+
+Example: `EDGE_VERSION='100.0.1185.29-1'`
+
+[Edge versions](https://packages.microsoft.com/repos/edge/pool/main/m/microsoft-edge-stable/)
+
+## Usage
```

Comment:
I guess all this assumes that there is already a `cypress/factory` image in the user's registry. How would a user build and tag `cypress/factory` locally? I tried `docker build . -f ./factory.Dockerfile -t cypress/factory` but got:

```
Step 1/21 : ARG BASE_IMAGE
Step 2/21 : FROM ${BASE_IMAGE} as factory
base name (${BASE_IMAGE}) should not be blank
```

...so I think I'm missing something. Could go at the end of this `Usage` section.

Reply:
We'll be publishing cypress/factory to dockerhub so people will have access to it. I don't really expect people to build it themselves.

for us we can run the following from the inside the `factory` directory.

`docker-compose build factory`

Reply:
It would be good to add that to this doc for future contributors. And also for me, because I didn't know how to test this locally :smile:

Reply:
Thats a good call.

Reply:
That command doesn't seem to work for me:

```
➜  factory git:(matth/docker-factory) docker-compose build factory
ERROR: The Compose file './docker-compose.yml' is invalid because:
services.factory.build contains unsupported option: 'tags'
services.chrome.build contains unsupported option: 'tags'
services.cypress-chrome.build contains unsupported option: 'tags'
services.cypress-edge.build contains unsupported option: 'tags'
services.cypress-firefox.build contains unsupported option: 'tags'
services.cypress.build contains unsupported option: 'tags'
services.edge.build contains unsupported option: 'tags'
services.firefox.build contains unsupported option: 'tags'
services.browsers.build contains unsupported option: 'tags'
services.base.build contains unsupported option: 'tags'
services.included.build contains unsupported option: 'tags'
```

```
➜  factory git:(matth/docker-factory) docker --version
Docker version 20.10.22, build 3a2c30b
➜  factory git:(matth/docker-factory) docker-compose --version
docker-compose version 1.29.2, build unknown
```

It says my Docker version should support `3.8` `docker-compose` files, any idea why it's not working?

Reply:
I had to update to the latest docker-compose version to get docker-compose to support the `tags` option

```
➜  factory git:(matth/docker-factory) docker --version
Docker version 20.10.21, build baeda1f
➜  factory git:(matth/docker-factory) docker-compose --version
Docker Compose version v2.13.0
```

Reply:
Thanks. Weird... Docker's default Ubuntu install has `docker-compose` 1.x, but if you install `docker-compose-plugin`, you can use `docker compose` 2.x...

```
➜  ~ docker compose version  
Docker Compose version v2.14.1
➜  ~ docker-compose --version
docker-compose version 1.29.2, build unknown
```

So with `docker compose` I am able to run that command to build the factory image.

Might also be worth mentioning that `docker-compose` 2.x is necessary because for some reason it seems to be elusive.

Reply:
thats good to know. I might go through and update any `docker-compose` commands to `docker compose` instead.

Reply:
Added instructions to build and test locally in the contributing document: https://github.com/cypress-io/cypress-docker-images/pull/812/commits/60088df3be182cee846520b9eb7be23de853d6f7